### PR TITLE
employ docker-compose v3.4 and shared logging config

### PR DESCRIPTION
**Pull request by fmigneault**
export only the specific changes of docker-compose version and logging config from #114 

Change overview:
* Cap container logs to 500m.  Before there was no limit to it could be logging indefinitively and eating up precious disk space.
* Remove unused `volume_from: mongodb` from `solr` and `ncwms2` container.
* Bump docker-compose file version to 3.4 to support sharing logging config.

Matching PR for Ouranos specific config https://github.com/bird-house/birdhouse-deploy-ouranos/pull/10

Fixes https://github.com/bird-house/birdhouse-deploy/issues/10

### CODE REVIEW

**Review by dbyrns — APPROVED** [review]
We'll wait for the tests to pass then it should be all good.

**Review by huard — APPROVED** [review]

File: birdhouse/docker-compose.yml
Comment (tlvu) [inline birdhouse/docker-compose.yml:7]: So we have not set any of those config currently (driver, max-size, max-file).  What's was their default?

It's very nice to have a unified logging config for all containers but can each container, if needed, not use this config and roll their own?

And to loop back to the original Weaver PR, Weaver need to "hook" into other components logs and that's why it prefers other components to have unified logging config or is there another reason to introduce this unified logging?

Just to be clear, I like it to have a default and unified logging config, I just want to understand its relation to Weaver.

File: birdhouse/docker-compose.yml
Comment (tlvu) [inline birdhouse/docker-compose.yml:217]: From what I understood from our meeting, the removal of `volumes_from` directive was because of the update to format version `3.4`, not because we want it or Weaver needs it?

File: docs/.gitignore
Comment (tlvu) [inline docs/.gitignore:2]: Not sure why you need to ignore everything under this folder by default.

I prefer the normal way to explicitely specify what you ignore to avoid ignoring "too much" and forgot to commit new files.

**Review by tlvu — COMMENTED** [review]
A few questions.  Not approving immediately because I'd like to try it on my dev server to see the difference with the new logging config but looks good overral.

File: birdhouse/docker-compose.yml
Comment (fmigneault) [reply birdhouse/docker-compose.yml:7]: Docker's default is json so that remains the same.
The difference here is that we add log rotation over 10 files. 
This is a config we have found a long time ago when deploying on our instances since twitcher and magpie can generate an immense quantity of logs (on each request). This allows discarding them gradually instead of accumulating GB of useless logs. 
Each container can define its own logging section by using the relevant config instead of the `default-logging` reference. 
I don't think there is a nice way to automate that though. Best could be to have an optional-component, but I think it is better to control the logs.

Comment (fmigneault) [reply birdhouse/docker-compose.yml:7]: Using Weaver, it is a bit more tricky since a different docker-compose file is employed. 
Depending on how docker-compose merges the files, the reference could or won't work across files. 
Basically, if files are merged, and then references resolved, it would work. But if internally it uses PyYAML or similar lib, to read each file individually and merge them, Weaver's compose will raise a missing reference. So the config would need to be duplicated in its own file.

File: docs/.gitignore
Comment (fmigneault) [reply docs/.gitignore:2]: I'm not sure if this is a remnant of an older commit build of the docs, but I had a lot of tmp doc files all over that directory (as in dozens of files). Running `make clean` removes them, but there is still a chance one could do `git add docs` and have lots of garbage collected. I find it is safe to do this kind of include as it is expected all doc files should be put in `sources` directory.

Comment (tlvu) [reply docs/.gitignore:2]: That's odd, what make command generate those cruft initially?  Maybe should fix the source of the problem instead (make all the temporary output under some dir and only add that specific dir to `.gitignore`.

Oh, I found the answer to your problem: https://github.com/bird-house/birdhouse-deploy/blob/57a320c3583f804838e8c18aeb064266527a0bc9/docs/Makefile#L50-L51

https://github.com/bird-house/birdhouse-deploy/blob/57a320c3583f804838e8c18aeb064266527a0bc9/docs/Makefile#L8

So you should just add folder `_build` to the `.gitignore` instead of ignoring everything.

Comment (fmigneault) [reply docs/.gitignore:2]: I can't tell. I'm also expecting all of them to be in `_build` since this is what is defined in Makefile, but I did have lots of files there. Like I said, maybe its due to old run. Either way, I don't have an objection, I just used that because I didn't want to include all the files in .gitignore manually.

Comment (tlvu) [reply docs/.gitignore:2]: But you said `make clean` removes them so all the files must have been under `_build`.  If files outside of `_build` meaning they are not created by this repo so why should `.gitignore` of this repo take care of them?

So basically if a `.gitignore` is required, I think only `_build` should be in there.  A "ignore everything" rule is dangerous and in this case not appropriate since the repo should only handle files it creates itself.

Comment (fmigneault) [reply docs/.gitignore:2]: I don't feel the dangerous aspect of it since there shouldn't be anything there anyway. All docs should be placed in `source`. Anyway, I will apply more specific `_build` if you prefer that. Same result for me.

File: birdhouse/docker-compose.yml
Comment (tlvu) [reply birdhouse/docker-compose.yml:7]: So if I understand correctly, Weaver did not need this central logging and the log rotation to avoid accumulating infinite logs (which is indeed a nice thing).

From our recent "contributing guideline", this PR should not have been mixed with the Weaver PR to begin with.  Or if mixed in, it should have been "well explained" that it is not required by Weaver?

File: birdhouse/docker-compose.yml
Comment (tlvu) [reply birdhouse/docker-compose.yml:217]: Agreed, this looks harmless but should be tested to be sure.  Has you or Mat gone over the failed build to determine the error is not related to this?  Currently only the ESGF notebook should be intermittently failing because their server seems flaky but everything else should be green.

**Review by tlvu — APPROVED** [review]
Please confirm failing tests are okay.

Please let me do this merge since I need to since this merge with my other PR https://github.com/bird-house/birdhouse-deploy-ouranos/pull/10 else our prod won't even start.

@nikola-rados I am pretty sure you will also need to do something like my PR https://github.com/bird-house/birdhouse-deploy-ouranos/pull/10 above on your side.  I can also sync with you for this merge.  Let me know.

File: birdhouse/docker-compose.yml
Comment (fmigneault) [reply birdhouse/docker-compose.yml:7]: Yes, this is why it has been split by itself now that we updated the contribution guideline. 
Please consider that Weaver PR was started in January and review only started in March, both long before we even started discussions of contribution guidelines.

**Comment by fmigneault** [comment]
@tlvu 
I have no problem waiting for your merge on this. 
If anything, I prefer you do it so that it is validated against a full stack. 

I'm not too sure what was the problem on the last build, it was stopped for some reason. 
Manual trigger produced the following outputs: 

=> http://daccs-jenkins.crim.ca/job/DACCS-iac/401 
=> http://daccs-jenkins.crim.ca/job/PAVICS-e2e-workflow-tests/job/master/316

```
22:12:17  notebooks/hummingbird.ipynb ............                                 [  8%]
22:12:25  notebooks-auth/test_thredds.ipynb ..........                             [ 15%]
22:12:32  pavics-sdi-master/docs/source/notebooks/WCS_example.ipynb .......        [ 20%]
22:13:06  pavics-sdi-master/docs/source/notebooks/WFS_example.ipynb .FFFF.         [ 24%]
22:13:15  pavics-sdi-master/docs/source/notebooks/WMS_example.ipynb ........       [ 30%]
22:13:17  pavics-sdi-master/docs/source/notebooks/WPS_example.ipynb ..........     [ 37%]
22:13:31  pavics-sdi-master/docs/source/notebooks/esgf-dap.ipynb .                 [ 37%]
22:13:32  pavics-sdi-master/docs/source/notebooks/jupyter_extensions.ipynb .       [ 38%]
22:13:40  pavics-sdi-master/docs/source/notebooks/opendap.ipynb .......            [ 43%]
22:13:44  pavics-sdi-master/docs/source/notebooks/pavics_thredds.ipynb ....F       [ 46%]
22:16:52  pavics-sdi-master/docs/source/notebooks/regridding.ipynb ...FF.FF.FFF.FF [ 57%]
22:16:56  FFFFFF...FFFFF.                                                          [ 67%]
22:17:00  pavics-sdi-master/docs/source/notebooks/rendering.ipynb ....             [ 70%]
22:17:05  pavics-sdi-master/docs/source/notebooks/subsetting.ipynb ......          [ 74%]
22:17:17  finch-master/docs/source/notebooks/dap_subset.ipynb ..........           [ 81%]
22:17:24  finch-master/docs/source/notebooks/finch-usage.ipynb ......              [ 86%]
22:17:51  finch-master/docs/source/notebooks/subset.ipynb ...FF....FFFFF....FF     [100%]
```

**Comment by fmigneault** [comment]
Just to confirm, it reran the whole thing, one time targeting [master](https://github.com/bird-house/birdhouse-deploy) and the other [docker-compose-update](https://github.com/bird-house/birdhouse-deploy/tree/docker-compose-update)

master
http://daccs-jenkins.crim.ca/job/DACCS-iac/402
http://daccs-jenkins.crim.ca/job/PAVICS-e2e-workflow-tests/job/master/317

docker-compose-update
http://daccs-jenkins.crim.ca/job/DACCS-iac/403
http://daccs-jenkins.crim.ca/job/PAVICS-e2e-workflow-tests/job/master/318

Results are the same in both cases, as presented below. 
Previous run of above https://github.com/bird-house/birdhouse-deploy/pull/152#issuecomment-828898807 just hit a connection timeout with boreas.ouranos.ca that trickled down failing many tests.

```
22:43:28  notebooks/hummingbird.ipynb ............                                 [  8%]
22:43:35  notebooks-auth/test_thredds.ipynb ..........                             [ 15%]
22:43:51  pavics-sdi-master/docs/source/notebooks/WCS_example.ipynb .......        [ 20%]
22:44:00  pavics-sdi-master/docs/source/notebooks/WFS_example.ipynb ......         [ 24%]
22:44:08  pavics-sdi-master/docs/source/notebooks/WMS_example.ipynb ........       [ 30%]
22:44:10  pavics-sdi-master/docs/source/notebooks/WPS_example.ipynb ..........     [ 37%]
22:44:25  pavics-sdi-master/docs/source/notebooks/esgf-dap.ipynb .                 [ 37%]
22:44:25  pavics-sdi-master/docs/source/notebooks/jupyter_extensions.ipynb .       [ 38%]
22:44:31  pavics-sdi-master/docs/source/notebooks/opendap.ipynb .......            [ 43%]
22:44:35  pavics-sdi-master/docs/source/notebooks/pavics_thredds.ipynb ....F       [ 46%]
22:46:28  pavics-sdi-master/docs/source/notebooks/regridding.ipynb ............... [ 57%]
22:47:12  ...............                                                          [ 67%]
22:47:18  pavics-sdi-master/docs/source/notebooks/rendering.ipynb ....             [ 70%]
22:47:23  pavics-sdi-master/docs/source/notebooks/subsetting.ipynb ......          [ 74%]
22:47:34  finch-master/docs/source/notebooks/dap_subset.ipynb ..........           [ 81%]
22:47:42  finch-master/docs/source/notebooks/finch-usage.ipynb ......              [ 86%]
22:48:08  finch-master/docs/source/notebooks/subset.ipynb ...FF....FFFFF....FF     [100%]
```

**Comment by tlvu** [comment]
@nikola-rados I am pretty sure you will also need to do something like my PR for Ouranos custom config https://github.com/bird-house/birdhouse-deploy-ouranos/pull/10 your side.  Let me know once you are done so we can sync this merge.

**Comment by nikola-rados** [comment]
@tlvu I have updated our config.

**Comment by tlvu** [comment]
> @tlvu I have updated our config.

@nikola-rados Thanks so I'll merge this during the day tomorrow so you can go-live manually if required.

**Comment by tlvu** [comment]
Default logging retention should have been bumped to something more reasonable for prod.  I forgot to push the bump here so bumping in PR https://github.com/bird-house/birdhouse-deploy/pull/165

**Pull request merged**
